File: solutions/haskell/bob/1/src/Bob.hs
module Bob (responseFor) where

import Data.Char (toUpper)
import Data.String (IsString (fromString))
import qualified Data.Text as T

responseFor :: String -> String
responseFor input
  | question && yelling = "Calm down, I know what I'm doing!"
  | question = "Sure."
  | yelling = "Whoa, chill out!"
  | silence = "Fine. Be that way!"
  | otherwise = "Whatever."
  where
    xs = (T.unpack . T.strip . fromString) input
    question = not silence && last xs == '?'
    yelling = not silence && xs == map toUpper xs && any (\c -> toUpper c `elem` ['A' .. 'Z']) xs
    silence = length xs == (length . filter (\c -> c `elem` [' ', '\t'])) xs

File: solutions/ocaml/bob/1/bob.ml
let response_for input =
  let input = String.trim input in
  let question = String.ends_with ~suffix:"?" input in
  let shouting =
    String.uppercase_ascii input = input
    && String.lowercase_ascii input <> input
  in
  let silence = input = "" in

  match (shouting, question, silence) with
  | true, true, _ -> "Calm down, I know what I'm doing!"
  | true, _, _ -> "Whoa, chill out!"
  | _, true, _ -> "Sure."
  | _, _, true -> "Fine. Be that way!"
  | _, _, _ -> "Whatever."
